File: space_invaders_test.py
# Noor Ansari

import unittest
from space_invaders import Game

class TestSpaceInvaders(unittest.TestCase):

    def setUp(self):
        self.game = Game()

    def test_player_movement_left_right(self):
        original = self.game.player_x
        self.game.move_player_left()
        self.assertEqual(self.game.player_x, max(0, original - 1))
        self.game.move_player_right()
        self.assertEqual(self.game.player_x, min(7, original))

    def test_fire_bullet(self):
        self.game.fire_bullet()
        self.assertTrue(self.game.bullet_active)
        self.assertEqual(self.game.bullet_y, 6)
        self.assertEqual(self.game.bullet_x, self.game.player_x)

    def test_bullet_moves_up(self):
        self.game.fire_bullet()
        start_y = self.game.bullet_y
        self.game.update_bullet()
        self.assertEqual(self.game.bullet_y, start_y - 1)

    def test_bullet_deactivates_off_screen(self):
        self.game.fire_bullet()
        self.game.bullet_y = 0
        self.game.update_bullet()
        self.assertFalse(self.game.bullet_active)

    def test_bullet_hits_enemy(self):
        self.game.fire_bullet()
        self.game.bullet_x = self.game.enemy_x
        self.game.bullet_y = self.game.enemy_y + 1
        self.game.update_bullet()
        self.assertFalse(self.game.bullet_active)
        self.assertEqual(self.game.enemy_y, 0)

    def test_enemy_reaches_player(self):
        self.game.enemy_y = 6
        self.game.enemy_x = self.game.player_x
        self.game.update_enemy()
        self.assertTrue(self.game.game_over)

if __name__ == '__main__':
    unittest.main()

File: SpaceInvadersTest.java
// Noor Ansari

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

public class SpaceInvadersTest {
    SpaceInvaders game;

    @BeforeEach
    void setup() {
        game = new SpaceInvaders();
    }

    @Test
    void testInitialPlayerPosition() {
        assertEquals(4, game.getPlayerX());
    }

    @Test
    void testMoveLeft() {
        game.movePlayerLeft();
        assertEquals(3, game.getPlayerX());
    }

    @Test
    void testMoveRight() {
        game.movePlayerRight();
        assertEquals(5, game.getPlayerX());
    }

    @Test
    void testFireBullet() {
        game.fireBullet();
        assertTrue(game.isBulletActive());
        assertEquals(game.getPlayerX(), game.getBulletX());
        assertEquals(6, game.getBulletY());
    }

    @Test
    void testBulletMovesUp() {
        game.fireBullet();
        int startY = game.getBulletY();
        game.updateBullet();
        assertEquals(startY - 1, game.getBulletY());
    }

    @Test
    void testBulletDeactivatesOffscreen() {
        game.fireBullet();
        for (int i = 0; i < 10; i++) {
            game.updateBullet();
        }
        assertFalse(game.isBulletActive());
    }

    @Test
    void testBulletHitsEnemy() {
        game.fireBullet();
        game.updateBullet(); // bulletY = 5
        game.updateBullet(); // bulletY = 4
        game.updateBullet(); // etc.

        // manual
        game.fireBullet();
        game.updateBullet();
        game.updateBullet();
        game.updateBullet();
        game.updateBullet();
        game.updateBullet();

        game.fireBullet();
        game.updateBullet();

        // set manually to hit
        game.fireBullet();
        game.updateBullet();
        game.updateBullet();
        game.updateBullet();
        game.updateBullet();
        game.updateBullet();
        game.updateBullet();
    }

    @Test
    void testEnemyReachesPlayer() {
        while (!game.isGameOver()) {
            game.updateEnemy();
            if (game.getEnemyY() == 7 && game.getEnemyX() == game.getPlayerX()) {
                break;
            }
        }
        assertTrue(game.isGameOver());
    }
}
